langlib/lang.value/src/main/java/org/ballerinalang/langlib/value/FromJsonWithType.java

```diff
@@ -0,0 +1,291 @@
+            case TypeTags.RECORD_TYPE_TAG:
+                BRecordType recordType = (BRecordType) targetType;
+                MapValueImpl<BString, Object> newRecord;
+                if (t != null && t.getDescribingType() == targetType) {
+                    newRecord = (MapValueImpl<BString, Object>) t.instantiate(strand);
+                } else {
+                    newRecord = (MapValueImpl<BString, Object>) BallerinaValues
+                            .createRecordValue(recordType.getPackage(), recordType.getName());
+                }
+
+                BType restFieldType = recordType.restFieldType;
+                Map<String, BType> targetTypeField = new HashMap<>();
+                for (BField field : recordType.getFields().values()) {
+                    targetTypeField.put(field.getFieldName(), field.getFieldType());
+                }
+
+                for (Map.Entry entry : map.entrySet()) {
+                    BType fieldType = targetTypeField.getOrDefault(entry.getKey().toString(), restFieldType);
+                    putToMap(newRecord, entry, fieldType, unresolvedValues, t, strand);
+                }
+                return newRecord;
+            case TypeTags.JSON_TAG:
+                BType matchingType = TypeConverter.resolveMatchingTypeForUnion(map, targetType);
+                return convert(map, matchingType, unresolvedValues, t, strand);
+            default:
+                break;
+        }
+        // should never reach here
+        throw BallerinaErrors.createConversionError(map, targetType);
+    }
+
+
+    private static Object convertArray(ArrayValue array, BType targetType, List<TypeValuePair> unresolvedValues,
+                                       TypedescValue t, Strand strand) {
+        switch (targetType.getTag()) {
+            case TypeTags.ARRAY_TAG:
+                BArrayType arrayType = (BArrayType) targetType;
+                ArrayValueImpl newArray = new ArrayValueImpl(arrayType);
+                for (int i = 0; i < array.size(); i++) {
+                    Object newValue = convert(array.get(i), arrayType.getElementType(), unresolvedValues, t, strand);
+                    newArray.add(i, newValue);
+                }
+                return newArray;
+            case TypeTags.TUPLE_TAG:
+                BTupleType tupleType = (BTupleType) targetType;
+                TupleValueImpl newTuple = new TupleValueImpl(tupleType);
+                int minLen = tupleType.getTupleTypes().size();
+                for (int i = 0; i < array.size(); i++) {
+                    BType elementType = (i < minLen) ? tupleType.getTupleTypes().get(i) : tupleType.getRestType();
+                    Object newValue = convert(array.get(i), elementType, unresolvedValues, t, strand);
+                    newTuple.add(i, newValue);
+                }
+                return newTuple;
+            case TypeTags.JSON_TAG:
+                newArray = new ArrayValueImpl(new BArrayType(BTypes.typeJSON));
```

Comment:
Since this array type is not going to change, how about adding it to `BTypes`, instead of creating a new JSON array type each time?